File: core/server/project/shine/src/main/java/com/home/shine/dataEx/ExternBuf.java
package com.home.shine.dataEx;

import com.home.shine.ctrl.Ctrl;
import com.home.shine.global.ShineSetting;

/** 外接buf */
public class ExternBuf
{
	public float[] fArr=new float[ShineSetting.externBufSize];
	public int[] iArr=new int[ShineSetting.externBufSize];
	public long[] lArr=new long[ShineSetting.externBufSize];
	
	private int _fIndex=0;
	private int _iIndex=0;
	private int _lIndex=0;
	
	public void clear()
	{
		_fIndex=0;
		_iIndex=0;
		_lIndex=0;
	}
	
	public void writeFloat(float v)
	{
		fArr[_fIndex++]=v;
	}
	
	public void writeInt(int v)
	{
		iArr[_iIndex++]=v;
	}
	
	public void writeLong(long v)
	{
		lArr[_lIndex++]=v;
	}
	
	public float readFloat()
	{
		return fArr[_fIndex++];
	}
	
	public int readInt()
	{
		return iArr[_iIndex++];
	}
	
	public long readLong()
	{
		return lArr[_lIndex++];
	}
}

File: core/server/projectC/shine/src/dataEx/ExternBuf.cpp
#include "ExternBuf.h"

ExternBuf::ExternBuf(float* fArr, jint* iArr, int64* lArr)
{
	this->fArr = fArr;
	this->iArr = iArr;
	this->lArr = lArr;
}

ExternBuf::~ExternBuf()
{
	this->fArr = nullptr;
	this->iArr = nullptr;
	this->lArr = nullptr;
}

void ExternBuf::clear()
{
	_fIndex = 0;
	_iIndex = 0;
	_lIndex = 0;
}

void ExternBuf::writeFloat(float v)
{
	fArr[_fIndex++] = v;
}

void ExternBuf::writeInt(int v)
{
	iArr[_iIndex++] = v;
}

float ExternBuf::readFloat()
{
	return fArr[_fIndex++];
}

void ExternBuf::writeLong(int64 v)
{
	lArr[_lIndex++] = v;
}

int ExternBuf::readInt()
{
	return iArr[_iIndex++];
}

int64 ExternBuf::readLong()
{
	return lArr[_lIndex++];
}

void ExternBuf::readVector(float* arr)
{
	arr[0] = readFloat();
	arr[1] = readFloat();
	arr[2] = readFloat();
}

void ExternBuf::writeVector(float* arr)
{
	writeFloat(arr[0]);
	writeFloat(arr[1]);
	writeFloat(arr[2]);
}

void ExternBuf::writeBoolean(bool v)
{
	writeInt(v ? 1 : 0);
}

bool ExternBuf::readBoolean()
{
	return readInt() == 1;
}
